server/update-workflow-interface/update-workflow-interface.md

```diff
@@ -0,0 +1,217 @@
+patterns for identifying a workflow execution and carrying arbitrary metadata
+(Header) and payloads (Payloads). The request type for this function includes a
+`wait_timeout` property indicating how long the client is willing to wait for a
+synchronous response once the request has been accepted by the system. This
+value can be set to zero indicating no such waiting should occur.  The
+`request_id` property uniquely identifies the logical request (_not_ the
+issuance of the request) and is intended for use as an idempotency token.
+
+The second function, `AwaitUpdateWorkflowResponse` is used when a client has
+already issued an `UpdateWorkflow` request and has the resulting `update_id`.
+That `update_id` can be provided along with a `wait_timeout` (same semantics as
+above) to re-attach to an in-flight or completed Update and obtain the
+associated UpdateResult in the form of an `UpdateWorkflowResponse`.
+
+
+```protobuf
+service WorkflowService {
+    rpc UpdateWorkflow(UpdateWorkflowRequest) returns (UpdateWorkflowResponse){}
+    rpc AwaitUpdateWorkflowResponse(AwaitUpdateWorkflowResponseRequest) returns (UpdateWorkflowResponse){}
+}
+
+message UpdateWorkflowRequest {
+    temporal.api.common.v1.Header header = 1;
+    string request_id = 2;
+    string namespace = 3;
+    temporal.api.common.v1.WorkflowExecution execution = 4;
+    string update_name = 5;
+    temporal.api.common.v1.Payloads input = 6;
+    google.protobuf.Duration wait_timeout = 7;
+}
+
+message UpdateWorkflowResponse {
+    UpdateResultType result_type = 1;
+    string update_id = 2;
+    temporal.api.common.v1.Payloads output = 3;
+}
+
+enum UpdateResultType {
+    UPDATE_RESULT_TYPE_UNSPECIFIED = 0;
+    UPDATE_RESULT_TYPE_SUCCESS = 1;
+    UPDATE_RESULT_TYPE_ERROR = 2;
+}
+
+message AwaitUpdateResponseRequest {
+    string update_id = 1; 
+    google.protobuf.Duration wait_timeout = 2;
+}
+```
+
+## Revisiting Use Cases With Propsed API
+
+### Short Update
+
+Client calls `UpdateWorkflowRequest` with a suitable `wait_timeout`.
+
+### Uncertain Duration Update
+
+Client calls `UpdateWorkflowRequest` with an acceptable `wait_timeout`. If the
+`result_type` of the returned `UpdateWorkflowResponse` is
+`UPDATE_RESULT_TYPE_UNSPECIFIED` then the client can infer that the Update is
+still runing and can use the returned `update_id` to obtain the UpdateResponse
+at a later time with `AwaitUpdateWorkflowResponse`. If the `result_type` is any
+other value then the response `output` can be used directly.
+
+### Known Long Update
+
+Client calls `UpdateWorkflowRequest` with a zero `wait_timeout` and uses the
+returned `update_id` to obtain the UpdateResponse at a later time with
+`AwaitWorkflowResponse`.
+
+### Invalid Update
+
+If the UpdateRequest is rejected by the workflow execution while the RPC client
+is still connected, a suitable gRPC error can be returned synchronously. Note
+however rejected UpdateRequests are _not_ recorded in workflow history.
```

Comment:
> rejected UpdateRequests are _not_ recorded in workflow history.

This is interesting. So what if I send an fire-and-forget update or short update to a "running" workflow that does not have a worker running it at the time? If I don't wait around it may not get in history but if I do wait around it does?

IMO like signals, a successfully received update by the _server_ should be all that is required for history persistence. The state of the worker should have nothing to do with whether the update will be processed in the future. It is every confusing to the caller that something as inexact as whether a worker is actively processing tasks for this workflow to affect some so exact as its history.

Reply:
> The update won't hit history until it's validated by the workflow so you can wait but until there's a worker running the update won't hit history.

But why? Is there real benefit to running worker code inline on every update before it hits history? I like the idea of sending an update like a signal.

> But the conclusion we came to was that the threat of filling history with invalid updates was a greater concern than keeping that consistency.

There is a threat of invalid updates but not invalid signals? Can we expand on this threat a bit? I don't think this is a real problem anymore than with signals. If updates were just signals with responses, you get so much more freedom. But forcing a worker to be involved in the creation of the update is rough IMO with very limited benefit and preventing a problem I am not sure exists.

Reply:
:-( I think that choice unfortunately hampers legitimate use of "always process the update regardless of worker state".

Reply:
I am going to close this thread as it's basically the same as the other.